packages/core/src/hydration/error_handling.ts

```diff
@@ -6,39 +6,366 @@
+  if (!node!.nextSibling) {
+    const header = 'During hydration Angular expected more sibling nodes to be present.\n\n';
+    const actual = `Actual DOM is:\n\n${describeActualDom(node!)}\n\n`;
+    const footer = getHydrationErrorFooter();
+
+    const message = header + actual + footer;
+    throw new RuntimeError(RuntimeErrorCode.HYDRATION_MISSING_SIBLINGS, message);
   }
 }
 
+/**
+ * Validates that a node exists or throws
+ */
 export function validateNodeExists(node: RNode|null): void {
   if (!node) {
-    // TODO: improve error message and use RuntimeError instead.
-    throw new Error(`Hydration expected an element to be present at this location.`);
+    throw new RuntimeError(
+        RuntimeErrorCode.HYDRATION_MISSING_NODE,
+        `Hydration expected an element to be present at this location.`);
+  }
+}
+
+/**
+ * Builds the hydration error message when a node is not found
+ *
+ * @param lView the LView where the node exists
+ * @param tNode the TNode
+ */
+export function nodeNotFoundError(lView: LView, tNode: TNode): Error {
+  const header = 'During serialization, Angular was unable to find an element in the DOM:\n\n';
+  const expected = `${describeExpectedDom(lView, tNode, false)}\n\n`;
+  const footer = getHydrationErrorFooter();
+
+  throw new RuntimeError(RuntimeErrorCode.HYDRATION_MISSING_NODE, header + expected + footer);
+}
+
+/**
+ * Builds a hydration error message when a node is not found at a path location
+ *
+ * @param host the Host Node
+ * @param path the path to the node
+ */
+export function nodeNotFoundAtPathError(host: Node, path: NodeNavigationStep[]): Error {
+  const header = `During hydration Angular was unable to locate a node ` +
+      `using the "${stringifyPath(path)}" path, starting from the ${describeRNode(host)} node.\n\n`;
+  const footer = getHydrationErrorFooter();
+
+  throw new RuntimeError(RuntimeErrorCode.HYDRATION_MISSING_NODE, header + footer);
+}
+
+
+// Stringification methods
+
+/**
+ * Stringifies a given TNode's attributes
+ *
+ * @param tNode a provided TNode
+ * @returns string
+ */
+function stringifyTNodeAttrs(tNode: TNode): string {
+  const results = [];
+  if (tNode.attrs) {
+    for (let i = 0; i < tNode.attrs.length;) {
+      const attrName = tNode.attrs[i++];
+      // Once we reach the first flag, we know that the list of
+      // attributes is over.
+      if (typeof attrName == 'number') {
+        break;
+      }
+      const attrValue = tNode.attrs[i++];
+      results.push(`${attrName}="${shorten(attrValue as string)}"`);
+    }
+  }
+  return results.join(' ');
+}
+
+/**
+ * The list of internal attributes that should be filtered out while
+ * producing an error message.
+ */
+const internalAttrs = new Set(['ngh', 'ng-version', 'ng-server-context']);
+
+/**
+ * Stringifies an HTML Element's attributes
+ *
+ * @param node an HTML Element
+ * @returns string
+ */
+function stringifyNodeAttrs(node: HTMLElement): string {
```

Comment:
(just for bigger contrast with `stringifyTNodeAttrs`)
```suggestion
function stringifyRNodeAttrs(node: HTMLElement): string {
```